src/mongoc/mongoc-topology-scanner.c

```diff
@@ -47,11 +48,31 @@ mongoc_topology_scanner_new (mongoc_topology_scanner_cb_t cb,
    BSON_APPEND_INT32 (&ts->ismaster_cmd, "isMaster", 1);
 
    ts->cb = cb;
-   ts->cb_data = cb_data;
+   ts->cb_data = data;
+   ts->uri = uri;
```

Comment:
you need to mongoc_uri_copy() the uri, otherwise it'll be destroyed by mongoc_client_new() which calls mongoc_topology_new() then mongoc_topology_scanner_new()